django_comments/views/comments.py

```diff
@@ -27,105 +33,157 @@ def __init__(self, why):
+        if self.request.user.is_authenticated:
+            comment.user = self.request.user
+
+        # Signal that the comment is about to be saved
+        responses = signals.comment_will_be_posted.send(
+            sender=comment.__class__,
+            comment=comment,
+            request=self.request
         )
 
-    # Otherwise create the comment
-    comment = form.get_comment_object(site_id=get_current_site(request).id)
-    comment.ip_address = request.META.get("REMOTE_ADDR", None) or None
-    if request.user.is_authenticated:
-        comment.user = request.user
-
-    # Signal that the comment is about to be saved
-    responses = signals.comment_will_be_posted.send(
-        sender=comment.__class__,
-        comment=comment,
-        request=request
-    )
-
-    for (receiver, response) in responses:
-        if response is False:
+        for (receiver, response) in responses:
+            if response is False:
+                raise BadRequest("comment_will_be_posted receiver %r killed the comment" % receiver.__name__)
+
+        # Save the comment and signal that it was saved
+        comment.save()
+        signals.comment_was_posted.send(
+            sender=comment.__class__,
+            comment=comment,
+            request=self.request
+        )
+        return comment
+
+    def get_template_names(self):
+        if self.template_name is None:
+            model = type(self.target_object)
+            return [
+                # These first two exist for purely historical reasons.
+                # Django v1.0 and v1.1 allowed the underscore format for
+                # preview templates, so we have to preserve that format.
+                "comments/%s_%s_preview.html" % (model._meta.app_label, model._meta.model_name),
+                "comments/%s_preview.html" % model._meta.app_label,
+                # Now the usual directory based template hierarchy.
+                "comments/%s/%s/preview.html" % (model._meta.app_label, model._meta.model_name),
+                "comments/%s/preview.html" % model._meta.app_label,
+                "comments/preview.html",
+            ]
+        else:
+            return [self.template_name]
+
+    def get_context_data(self, form):
+        return dict(
+            form=form,
+            comment=form.data.get("comment", ""),
+            next=self.data.get("next", self.kwargs.get('next')),
+        )
+
+    @method_decorator(csrf_protect)
+    def dispatch(self, *args, **kwargs):
+        return super().dispatch(*args, **kwargs)
+
+    def post(self, request, **kwargs):
+        self.object = None
+        self.target_object = None
+        self.data = self.get_form_kwargs()
+        try:
+            self.target_object = self.get_target_object(self.data)
+        except BadRequest as exc:
+            return exc.response
+
+        form = self.get_form()
+
+        # Check security information
+        if form.security_errors():
```

Comment:
Please explain what is your intention here. From what I can see you shouldn't need to override the post method and not call `get_form_kwargs()` or do form validation manually. All of that is supposed to be handled by FormView under the hood.

Reply:
As I expained above the post method is the first one called. It's the best place to set object attributes that are needed throught out the code, like self.data and self.target_object.

Reply:
In the inherited post method only two cases are differentiated: form-valid and form-invalid. In the logic of the original post_comment view there is a third one: form.security_errors(). Another reason to customize the post-method.

Reply:
I see the `post` method here as the original `post_comment` method. The original content of the `post_comment` is split in other methods inherited from the `FormView` and what is not in other methods remain here in the `post` method. I think it's a good approach.

The reason why I believe the call to `form.security_errors` is here, is because the `is_valid` logic of the `CommentForm` does not do it. Putting it somewhere else and leaving the class without this `post` method would leave this new implementation a bit less readable. It could be in the `CommentForm`, but it would represent more side effects to look after.

What do you think? What do you suggest to do to continue moving forward with this PR? I mean I really like what @thomst did here and it would be great if it landed in the repository. 
Thanks for your input :-)